vosekast_control/TestSequence.py

```diff
@@ -31,134 +32,205 @@ def __init__(
+                    # await constant_tank full
+                    await self.vosekast.constant_tank.fill()
+
+                # check ready_to_measure
+                if not self.vosekast.ready_to_measure:
+                    self.logger.debug("Vosekast not ready to measure.")
+                    self.scale.print_diagnostics()
+                    return
+                # check state
+                elif self.state == self.STOPPED or self.state == self.PAUSED:
+                    return
+
+                # turn on measuring pump, start measuring
+                await self.start_measuring()
+
+                self.vosekast.state = self.vosekast.MEASURING
+                self.state = self.MEASURING
+
+                # write to file
+                await self.write_loop()
+
+            # TankFillingTimeout
+            except:
+                self.logger.error("Error, aborting test sequence.")
+
+                await self.stop_sequence()
+                self.vosekast.constant_tank.state = self.vosekast.constant_tank.STOPPED
+                self.vosekast.state = self.vosekast.RUNNING
+        else:
+            self.logger.info("Already measuring.")
 
-            await self.vosekast.constant_tank.fill()
-
-            # check
-            if not self.vosekast.ready_to_measure():
-                self.logger.debug("Vosekast not ready to measure.")
-                self.scale.print_diagnostics()
-                return
-
-            # turn on measuring pump, start measuring
-            await self.start_measuring()
-
-            self.vosekast.state = "MEASURING"
-
-            # write to file
-            await self.write_loop()
-
-        # TankFillingTimeout
-        except:
-            self.logger.error("Error, aborting test sequence.")
+    async def write_loop(self):
+        if not self.emulate:
+            try:
+                # null scale
+                if abs(self.scale.scale_history[0]) < 0.15:
+                    scale_nulled = self.scale.scale_history[0]
+                else:
+                    scale_nulled = 0
+
+                while self.state == self.MEASURING and not self.vosekast.measuring_tank.is_filled:
+                    # get flow average
+                    flow_average = self.scale.flow_average()
+                    scale_actual = round(
+                        self.scale.scale_history[0] - scale_nulled, 5)
+
+                    # #write values to csv file
+                    # with open('sequence_values.csv', 'a', newline='') as file:
+                    #     writer = csv.writer(file, delimiter=',',
+                    #                     quotechar='|', quoting=csv.QUOTE_MINIMAL)
+                    #     writer.writerow([self.scale.scale_history[1], scale_actual, self.scale.flow_history[0], flow_average])
+
+                    # db
+                    dbconnect = sqlite3.connect('sequence_values.db')
```

Comment:
please add a todo comment here, that we have to move the db init in a class or a singleton

Reply:
ok